Code/Framework/AzCore/AzCore/Debug/Trace.h

```diff
@@ -39,14 +37,122 @@ namespace AZ
             None
         };
 
-        class Trace
+        class ITrace
         {
         public:
-            static Trace& Instance()    { return g_tracer; }
+            O3DEKERNEL_API ITrace()
+            {
+                tracer = this;
+            }
+            O3DEKERNEL_API virtual ~ITrace()
+            {
+                tracer = nullptr;
+            }
+            ITrace(const ITrace&) = delete;
+            O3DEKERNEL_API ITrace(ITrace&&) = default;
+            ITrace& operator=(const ITrace&) = delete;
+            O3DEKERNEL_API ITrace& operator=(ITrace&&) = default;
+
+            O3DEKERNEL_API static ITrace& Instance()
+            {
+                if (!tracer)
+                {
+                    static ITrace defaultTracer;
+                }
+                return *tracer;
+            }
+
+            O3DEKERNEL_API static void SetInstance(ITrace* newTracer)
+            {
+                tracer = newTracer;
+            }
 
+            O3DEKERNEL_API virtual void Init() {}
+            O3DEKERNEL_API virtual void Destroy() {}
+            O3DEKERNEL_API virtual bool IsDebuggerPresent() { return false; }
+            O3DEKERNEL_API virtual void Break() {}
+            O3DEKERNEL_API virtual void Crash() {}
+
+            /// Indicates if trace logging functions are enabled based on compile mode and cvar logging level
+            O3DEKERNEL_API bool IsTraceLoggingEnabledForLevel(LogLevel level)
+            {
+                return m_logLevel >= level;
+            }
+            O3DEKERNEL_API void SetLogLevel(LogLevel newLevel)
+            {
+                m_logLevel = newLevel;
+            }
+
+            O3DEKERNEL_API bool AlwaysPrintCallstack()
+            {
+                return m_printCallstack;
+            }
```

Comment:
How about `GetAlwaysPrintCallstack()`?

My instant reaction when I saw this function, was that I thought it was actually printing the callstack.
```suggestion
            O3DEKERNEL_API bool GetAlwaysPrintCallstack()
            {
                return m_printCallstack;
            }
```

Reply:
Also this function should be marked const.